src/components/3d_editor/advanced_geometry/PythonTransformation.jsx

```diff
@@ -0,0 +1,352 @@
+            pyodideInstance.setStdout({ batched: (text) => this.handleStdout(text) });
+            document.pyodideMplTarget = document.getElementById("pyodide-plot-target");
+        });
+    };
+
+    loadPackages = async () => {
+        const { pyodide } = this.state;
+
+        if (pyodide) {
+            await pyodide.runPythonAsync(installPkg);
+        }
+        this.setState({ isLoading: false });
+    };
+
+    loadPythonCode = () => {
+        const { transformationParameters } = this.state;
+        const code = transformationsMap[transformationParameters.transformationKey].content;
+        this.setState({ pythonCode: code });
+    };
+
+    handleCodeChange = (newContent) => {
+        this.setState({ pythonCode: newContent });
+    };
+
+    runPythonCode = async () => {
+        const dataOut = null;
+        const { pyodide, pythonCode } = this.state;
+        this.setState({ pythonOutput: "" });
+        document.pyodideMplTarget.innerHTML = "";
+
+        try {
+            const result = await pyodide.runPythonAsync(pythonCode);
+            console.log(result);
+        } catch (error) {
+            this.setState({ pythonOutput: error.message });
+        }
+        return dataOut;
+    };
+
+    handleRun = async () => {
+        this.setState({ isRunning: true });
+        try {
+            await this.runPythonCode();
+        } catch (error) {
+            NPMsAlert.error(error.message);
+        } finally {
+            this.setState({ isRunning: false });
+        }
+    };
+
+    handleSubmit = () => {
+        const { onSubmit } = this.props;
+        const { newMaterials } = this.state;
+        onSubmit(newMaterials);
+    };
+
+    handleTransformationParametersChange = (event, newValue) => {
+        if (newValue) {
+            const { transformationParameters } = this.state;
+            this.setState({
+                transformationParameters: {
+                    ...transformationParameters,
+                    transformation: newValue,
+                },
+            });
+            this.loadPythonCode();
+        }
+    };
+
+    handleMaterialSelectionChange = (event, newValue) => {
+        this.setState({ selectedMaterials: newValue });
+    };
+
+    render() {
+        const {
+            theme,
+            isLoading,
+            isRunning,
+            pythonCode,
+            pythonOutput,
+            transformationParameters,
+            materials,
+            selectedMaterials,
+        } = this.state;
+        const codemirrorTheme = theme === LightMaterialUITheme ? "light" : "dark";
+        const { show, onHide } = this.props;
+
+        const getStatusText = () => {
+            if (isLoading) return "Loading...";
+            if (isRunning) return "Running...";
+            return "Ready";
+        };
+
+        const icon = <CheckBoxOutlineBlank fontSize="small" />;
+        const checkedIcon = <CheckBox fontSize="small" />;
+
+        const controls = () => {
```

Comment:
This would better be a separate component `PythonTransformationControls` or something like that

Reply:
Was thinking about that. If you find separate component to be more readable -- I'll do it.